plugins/nvme/fu-nvme-device.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fcntl.h>
+#include <string.h>
+#include <sys/errno.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <glib/gstdio.h>
+
+#include <efivar.h>
+#include <linux/nvme_ioctl.h>
+
+#include "fu-chunk.h"
+#include "fu-nvme-device.h"
+
+#define FU_NVME_ID_CTRL_SIZE	0x1000
+
+struct _FuNvmeDevice {
+	FuDevice		 parent_instance;
+	GUdevDevice		*udev_device;
+	guint			 pci_depth;
+	gint			 fd;
+};
+
+G_DEFINE_TYPE (FuNvmeDevice, fu_nvme_device, FU_TYPE_DEVICE)
+
+#ifndef HAVE_GUDEV_232
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-function"
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevDevice, g_object_unref)
+#pragma clang diagnostic pop
+#endif
+
+static void
+fu_nvme_device_to_string (FuDevice *device, GString *str)
+{
+	FuNvmeDevice *self = FU_NVME_DEVICE (device);
+	g_string_append (str, "  FuNvmeDevice:\n");
+	g_string_append_printf (str, "    fd:\t\t\t%i\n", self->fd);
+	g_string_append_printf (str, "    pci-depth:\t\t%u\n", self->pci_depth);
+}
+
+static gchar *
+fu_nvme_device_get_string_safe (const guint8 *buf, guint16 addr_start, guint16 addr_end)
+{
+	GString *str;
+
+	g_return_val_if_fail (buf != NULL, NULL);
+	g_return_val_if_fail (addr_start < addr_end, NULL);
+
+	str = g_string_new_len (NULL, addr_end + addr_start + 1);
+	for (guint16 i = addr_start; i <= addr_end; i++) {
+		gchar tmp = (gchar) buf[i];
+		/* skip leading spaces */
+		if (g_ascii_isspace (tmp) && str->len == 0)
+			continue;
+		if (g_ascii_isprint (tmp))
+			g_string_append_c (str, tmp);
+	}
+
+	/* nothing found */
+	if (str->len == 0) {
+		g_string_free (str, TRUE);
+		return NULL;
+	}
+	return g_strchomp (g_string_free (str, FALSE));
+}
+
+static gchar *
+fu_nvme_device_get_guid_safe (const guint8 *buf, guint16 addr_start)
+{
+	efi_guid_t guid_tmp;
+	guint c_cnt = 0;
+	g_autofree char *guid = NULL;
+
+	/* check the GUID is plausible */
+	for (guint i = 0; i < 16; i++)
+		c_cnt += buf[addr_start + i];
+	if (c_cnt < 0xff)
```

Comment:
Is this really true?  all EFI guids add up to something smaller than 255?  I'd never heard that, but it's great to know if it is.

Reply:
Nope, that's completely a made up rule. I figured statistically the GUID would have an average summed value of 2032, and that GUIDs like `01000200-0000-0300-0010-080020000a00` would be statistically unlikely and it probably indicates it's not actually a GUID.

Reply:
Sounds pretty good.
What would you think about putting in a warning for this situation that it found some data but that it doesn't look like a GUID?  That way at least if we find this happening in the wild we can see warnings in fwupd.log to go with it.

Reply:
Well, as I understand it there are quite a few drives with component IDs and not GUIDs, so wouldn't a lot of drives have warnings? Are the early drives going to have zeros for the GUID? I guess we could avoid a warning for that.

Reply:
Yeah they should have 0's.

Reply:
Okay, fixed.